### ISSUE TITLE
SparseApply* operators for the GPU

### ISSUE THREAD

**Issue by stephenroller** [tensorflow/tensorflow#2314]
At the moment, it appears that only GradientDescentOptimizer supports running on the GPU when there is a SparseTensor update. This is particularly relevant for any RNNs which train their embeddings, including the word2vec example (yes w2v isn't an RNN ;)

So far the common workaround everyone seems to use is to force the embedding variables to be on the CPU, but there can be _substantial_ speed improvements by allowing them to be stored on the GPU. For one, there is no need to transfer the vectors of embeddings to/from the GPU, and instead one can just transfer the embedding indexes, and then the gradients also don't need to be transferred backwards. In one test I've run where I implemented a version on the GPU, the difference in one epoch was roughly 1100s vs 300s. Basically anyone who doesn't freeze their embeddings can substantially benefit from this.

This is related to #1310 and #464. In #1310, there is a bug where variables are placed on the GPU, even though an op that appears later isn't available on the GPU. My understanding is a fix for this is under development, but it will only make sure variables are placed on devices which can perform the necessary ops. It will not actually include SparseApply's. This feature request would resolve the OP's reported bug in #1310, but not solve the variable-operator placement issue.

In #464, the issue is that no one has implemented any `SparseApplyRMSProp`, along with another bug that has since been resolved. The issue used to contain all optimizers except GDO, but since then SparseApply ops have been added for AdaGrad, yet they're currently only implemented for the CPU, not the GPU. This feature request is a rough superset of #464.

One thing that is presently unclear to me, is whether these really need to be implemented as C++ operators, with all the associated book keeping. Currently that is how the CPU version of the SparseApply*'s are updated (see [example](https://github.com/tensorflow/tensorflow/blob/5681406e2874a02835d34be579810a93ad74a473/tensorflow/core/kernels/training_ops.cc#L1101)), except for [GradientDescent](https://github.com/tensorflow/tensorflow/blob/5681406e2874a02835d34be579810a93ad74a473/tensorflow/python/training/gradient_descent.py#L54). Its `_sparse_apply` is implemented in python and uses `scatter_sub`.

I already have a Python implementation of `_sparse_apply` for `MomentumOptimizer`, which uses `tf.gather`, `tf.scatter_update`, and `tf.scatter_sub`. It passes unit tests, and performs on both CPU and the GPU, though it does take about a 1% performance hit compared to the current CPU C++ `SparseApplyMomentum`. Would this be of interest, or is there a motivation behind implementing this in C++?

**Comment by girving** [comment]
Apologies, this seems to have fallen through the cracks (@mrry).  We'd be very interested in GPU versions of these ops, but it's unfortunate if we have to sacrifice CPU performance to get them.

@vrv: Can you think of a clean way to use a fused C++ op on CPU but a combination of separate ops on GPU?  I'd be tempted to just accept the fix, but people will probably yell at me if I let a performance regression through.

**Comment by vrv** [comment]
IIRC, the reason for implementing all of these as C++ ops is due to race conditions between read/compute/update.  If you implement these as separate ops, the time between computing the various compositions of the update rule is much higher than if it is implemented in a single op.  We found this caused problems for multi-replica models due to racy updates, which is why I think most of these ops are fused.  So I'd be a bit concerned about adding the composite op version without sufficient testing.

And no, I can't think of an easy way right now to do this, since we don't have a good rewrite infrastructure that's device dependent.

**Comment by stephenroller** [comment]
Yeah I didn't think about race conditions on multiple machines; that would be a problem. I also spent a few days trying to implement the operator in C++, but my C++ is relatively weak, I don't know CUDA at all. So that wasn't fruitful.

A week later I was talking to a colleague about this and I noticed that my patch probably also computes updates incorrectly if the same embedding is updated multiple times. Rather than adding all the gradients together, then updating the momentum, and then updating each var exactly once, it would repeatedly update momentum and the variable. As far as I can tell, the current CPU SparseApplyMomentum operator probably has the same bug, but I haven't tested anything; it's just thought experiments.

I also haven't looked at this in a few weeks, so I'd have to rebase the patch and make sure it's still fresh anyway. I'll do that as soon as I get a chance, sometime this week.

**Comment by girving** [comment]
@stephenroller: Cool, I'll leave this open as contributions welcome, but based on Vijay's comments it would have to be a unified op to reduce races.  The GPU scatter and gather kernels are reasonably clean, so it wouldn't be too hard to mimic them.

**Comment by liuyipei** [comment]
@stephenroller 
Sorry for resurrecting an old thread. I was just wondering if there are any updates on SparseApply* Issue.

**Comment by stephenroller** [comment]
Unfortunately my own research ended up taking me in a different direction, so I never got around to this, and probably won't in the immediate future. It would certainly still be welcome on my end though, if someone ever implements this.

**Comment by girving** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by drasmuss** [comment]
Does that mean that there are no plans to implement SparseApply operators for the GPU (or that they've already been added)?  Or is there another issue tracking that feature? I'd just been following this one waiting for updates.

**Comment by girving** [comment]
@drasmuss It's been contributions welcome since November, and the person who said they might do it went a different direction.  We'd be happy to accept pull requests, but we're closing issues where no activity is on the horizon.  Would you like to submit a PR?

**Comment by stephenroller** [comment]
As far as I can tell, this may have actually been addressed by
https://github.com/tensorflow/tensorflow/commit/67443722b26c3585d860d44e7069d997300a7187

**Comment by drasmuss** [comment]
Finally had a chance to check this, and just for anyone else coming across this thread, it looks like SparseApply* ops are still not implemented for the GPU (as of TF 1.14/2.0 beta) E.g., see https://github.com/tensorflow/tensorflow/blob/c174697c09737e543e1919ca5cd5cfc373a96012/tensorflow/core/kernels/training_ops.cc#L3646).

6744372 added support for SparseApply ops with ResourceVariables, but that's still CPU-only.

**Comment by sjtusmartboy** [comment]
is there any update of this issue, since I also terribly need the SparseApply* operators on gpu
